File: src/actor.mjs
import { MemoryError } from "./errors.mjs";
const AGENT = /^[A-Za-z0-9._-]+\/[A-Za-z0-9._:-]+$/;
const HUMAN = /^human:[A-Za-z0-9._-]+$/;
const PROCESS = /^process:[A-Za-z0-9._-]+$/;
export function parseActor(s) {
  if (typeof s !== "string") throw new MemoryError("refused", "actor must be a string");
  if (HUMAN.test(s)) return { value: s, class: "human" };
  if (PROCESS.test(s)) return { value: s, class: "process" };
  if (AGENT.test(s)) return { value: s, class: "agent" };
  throw new MemoryError("refused", `malformed actor ${JSON.stringify(s)}: use <producer>/<version>, human:<id> or process:<id>`);
}

File: memory-go/actor.go
package memory

import "regexp"

var (
	actorAgent   = regexp.MustCompile(`^[A-Za-z0-9._-]+/[A-Za-z0-9._:-]+$`)
	actorHuman   = regexp.MustCompile(`^human:[A-Za-z0-9._-]+$`)
	actorProcess = regexp.MustCompile(`^process:[A-Za-z0-9._-]+$`)
)

// Actor is OKF §7 identity: <producer>/<version>, human:<id> or process:<id>.
type Actor struct {
	Value string `json:"value"`
	Class string `json:"class"`
}

func ParseActor(s string) (Actor, error) {
	switch {
	case actorHuman.MatchString(s):
		return Actor{s, "human"}, nil
	case actorProcess.MatchString(s):
		return Actor{s, "process"}, nil
	case actorAgent.MatchString(s):
		return Actor{s, "agent"}, nil
	}
	return Actor{}, Errorf(CodeRefused, "malformed actor %s: use <producer>/<version>, human:<id> or process:<id>", quoteJSON(s))
}
